app/client/components/duplicateWidget.ts

```diff
@@ -0,0 +1,318 @@
+import {cleanFormLayoutSpec} from 'app/client/components/FormRenderer';
+import {GristDoc} from 'app/client/components/GristDoc';
+import {BoxSpec, purgeBoxSpec} from 'app/client/lib/BoxSpec';
+import {makeT} from 'app/client/lib/localization';
+import {ViewFieldRec, ViewSectionRec} from 'app/client/models/DocModel';
+import {cssField, cssLabel} from 'app/client/ui/MakeCopyMenu';
+import {IPageWidget, toPageWidget} from 'app/client/ui/PageWidgetPicker';
+import {select} from 'app/client/ui2018/menus';
+import {saveModal} from 'app/client/ui2018/modals';
+import {BulkColValues, getColValues, RowRecord, UserAction} from 'app/common/DocActions';
+import {arrayRepeat} from 'app/common/gutil';
+import {schema} from 'app/common/schema';
+import {dom, Observable} from 'grainjs';
+import cloneDeepWith from 'lodash/cloneDeepWith';
+import flatten from 'lodash/flatten';
+import forEach from 'lodash/forEach';
+import {testId} from 'app/client/ui2018/cssVars';
+import {logTelemetryEvent} from 'app/client/lib/telemetry';
+import sortBy from 'lodash/sortBy';
+import {fromPairs} from 'lodash';
+
+const t = makeT('duplicateWidget');
+
+// Duplicate page with pageId. Starts by prompting user for a new name.
+export async function buildDuplicateWidgetModal(gristDoc: GristDoc, viewSectionId: number) {
+  const activeView = gristDoc.activeViewId.get();
+  const pages = sortBy(gristDoc.docModel.pages.rowModels.filter(p => p), [(row) => row.pagePos.peek()]);
+  const pageSelectOptions = pages.map(page => {
+    const view = page.view.peek();
+    const isActivePage = view.getRowId() === activeView;
+    const suffix = isActivePage? " (Active)" : "";
+    return {
+      label: `${view.name.peek()}${suffix}`,
+      value: view.getRowId(),
+      isActivePage: isActivePage,
+    };
+  });
+
+  pageSelectOptions.push({ label: 'Create new page', value: 0, isActivePage: false });
+
+  // Logically this should never happen, as a Grist doc without pages should be impossible.
```

Comment:
It is impossible because you've just added an item above :)

Reply:
:rofl:, great catch, thanks Jarek :)